### ISSUE TITLE
Memory leak with Python 3.11.2

### ISSUE THREAD

**Issue by scotloach** [fafhrd91/aiohttp#7252]
### Describe the bug

I am seeing a memory leak with aiohttp using Python 3.11.2.

When I downgrade to Python 3.11.0, it does not leak.

This is a script that does a large number of client requests to a large number of different servers.  The script leaks hundreds of MB over the course of a few hours.  I haven't yet gotten this to a form that can be easily reproduced and shared, but I thought I would get this entered in case anyone else is seeing similar symptoms.

### To Reproduce

The script I am running creates about 700 https client sessions to different servers.  Each session does a simple GET request, which most succeed but some fail in various ways.  The script sleeps for 5 seconds and then repeats this.

### Expected behavior

This should not leak memory over time.  And it doesn't with 3.11.0.  (I don't currently have an easy way to run it with 3.11.1).

### Logs/tracebacks

```python-traceback
n/a
```

### Python Version

```console
Python 3.11.2
```

### aiohttp Version

```console
Name: aiohttp
Version: 3.8.4
Summary: Async http client/server framework (asyncio)
Home-page: https://github.com/aio-libs/aiohttp
Author: 
Author-email: 
License: Apache 2
Location: /usr/lib64/python3.11/site-packages
Requires: aiosignal, async-timeout, attrs, charset-normalizer, frozenlist, multidict, yarl
```

### multidict Version

```console
Name: multidict
Version: 6.0.4
Summary: multidict implementation
Home-page: https://github.com/aio-libs/multidict
Author: Andrew Svetlov
Author-email: andrew.svetlov@gmail.com
License: Apache 2
Location: /usr/lib64/python3.11/site-packages
Requires: 
Required-by: aiohttp, yarl
```

### yarl Version

```console
Name: yarl
Version: 1.8.2
Summary: Yet another URL library
Home-page: https://github.com/aio-libs/yarl/
Author: Andrew Svetlov
Author-email: andrew.svetlov@gmail.com
License: Apache 2
Location: /usr/local/lib64/python3.11/site-packages
Requires: idna, multidict
Required-by: aiohttp
```

### OS

Fedora release 37 (Thirty Seven)

### Related component

Client

### Code of Conduct

- [X] I agree to follow the aio-libs Code of Conduct

**Comment by Dreamsorcerer** [comment]
If the memory leak appears only in a specific patch version of Python, I suspect the bug report will need to go to cpython.
If you can profile the memory to atleast figure out which objects are causing the memory leak, that would certainly help figure out the cause.

**Comment by Dreamsorcerer** [comment]
Also, sharing the script would be useful.

**Comment by scotloach** [comment]
tracemalloc profile after running for awhile: (This is python 3.11.3 which also exhibits the leak)
`/usr/lib64/python3.11/asyncio/sslproto.py:274: size=349 MiB, count=2788, average=128 KiB
/usr/lib64/python3.11/asyncio/sslproto.py:374: size=14.7 MiB, count=241614, average=64 B
/usr/lib64/python3.11/asyncio/sslproto.py:343: size=2156 KiB, count=2788, average=792 B
/usr/lib64/python3.11/site-packages/aiohttp/connector.py:598: size=1505 KiB, count=112, average=13.4 KiB
/usr/lib64/python3.11/asyncio/sslproto.py:305: size=1035 KiB, count=2788, average=380 B
/usr/lib64/python3.11/site-packages/aiohttp/connector.py:666: size=1015 KiB, count=40, average=25.4 KiB
/usr/lib64/python3.11/site-packages/aiohttp/client_proto.py:213: size=525 KiB, count=7042, average=76 B
/usr/lib64/python3.11/site-packages/aiohttp/streams.py:571: size=507 KiB, count=1366, average=380 B
/usr/lib64/python3.11/site-packages/aiohttp/connector.py:365: size=467 KiB, count=42, average=11.1 KiB`

This doesn't accurately reflect the heap sizes I see, but I suspect that second line with 241614 allocations is the issue.
This is with long running ClientSession objects.  If I close the ClientSession objects and re-initialize them after each request, it doesn't leak.

This is the main body of the script I'm using to reproduce it; it's loading a set of around 700 private IP endpoints which respond in a variety of different ways.

```
def init():
    timeout = aiohttp.ClientTimeout(connect=5, sock_read=5)
    ssl_ctx = ssl._create_unverified_context()
    conn = aiohttp.TCPConnector(ssl_context=ssl_ctx, enable_cleanup_closed=True)
    session = aiohttp.ClientSession(connector=conn, timeout=timeout, cookie_jar=aiohttp.CookieJar(unsafe=True))
    return session
    
async def fetch(client, ip):
    try:
        async with client.request('GET', url=f'https://{ip}:8000/status.cgi') as r:
            msg = await r.text()
    except asyncio.CancelledError:
        raise
    except Exception as err:
        pass

async def main(): 
    ips = [...] # load IPs to poll here
    clients = {ip: init() for ip in ips}
    tracemalloc.start()
    try:
        while True:
            await asyncio.gather(*[fetch(client, ip) for ip, client in clients.items()])
            await asyncio.sleep(5)
    except asyncio.CancelledError:
        pass  # end and clean things up
    finally:
        memory_used = tracemalloc.get_traced_memory()
        snapshot = tracemalloc.take_snapshot()
        stats = snapshot.statistics('lineno')
        for stat in stats[:10]:
            print(stat)
        try:
            await asyncio.gather(*[client.close() for client in clients.values()])
        except:
            pass

asyncio.run(main()) 
```

**Comment by scotloach** [comment]
I have an update.  It seems that removing the `enable_cleanup_closed=True` from my test script, has resolved the leak.

I know I had to put that in at some earlier time to _prevent_ a memory leak.  I wonder if something has changed in recent Python versions that has fixed the cause of this, and now that flag ends up _causing_ a leak.

**Comment by Dreamsorcerer** [comment]
I'm not familiar with the code, but at a glance it just appends to a list:
https://github.com/aio-libs/aiohttp/blob/master/aiohttp/connector.py#L314
And create a weakref:
https://github.com/aio-libs/aiohttp/blob/master/aiohttp/connector.py#L363

Maybe check the size of that list? I think your tracealloc might be indicating something with that too. No idea why a cpython change would affect the behaviour though.

**Comment by bdraco** [comment]
I'm seeing the leak as well with `enable_cleanup_closed=True`

I see a lot of `_SSLProtocolTransport` objects in memory with no socket attached to them.

```
2023-05-12 23:34:58.325 CRITICAL (MainThread) [homeassistant.components.profiler] _SSLProtocolTransport <asyncio.sslproto._SSLProtocolTransport object at 0x7fa1382d6210> ssl_proto=None ssl_object=<ssl.SSLObject object at 0x7fa156521650> peercert={'su
bject': ((('commonName', 'new-home.sensibo.com'),),), 'issuer': ((('countryName', 'US'),), (('organizationName', 'Amazon'),), (('commonName', 'Amazon RSA 2048 M02'),)), 'version': 3, 'serialNumber': '0FFA0B652D4CB4ADA67C967B5136739D', 'notBefore': 'F
eb 23 00:00:00 2023 GMT', 'notAfter': 'Mar 23 23:59:59 2024 GMT', 'subjectAltName': (('DNS', 'new-home.sensibo.com'), ('DNS', 'home.sensibo.com')), 'OCSP': ('http://ocsp.r2m02.amazontrust.com',), 'caIssuers': ('http://crt.r2m02.amazontrust.com/r2m02.
cer',), 'crlDistributionPoints': ('http://crl.r2m02.amazontrust.com/r2m02.crl',)} sock=None
```

**Comment by bdraco** [comment]
https://github.com/python/cpython/issues/98539 looks related

**Comment by bdraco** [comment]
https://github.com/python/cpython/pull/98540

**Comment by bdraco** [comment]
related https://github.com/elastic/rally/pull/1712

**Comment by bdraco** [comment]
related https://github.com/elastic/rally/issues/1714

**Comment by webknjaz** [comment]
> related https://github.com/elastic/rally/pull/1712

Looks like this was backported via https://github.com/python/cpython/commit/bd8b32b519055b5411e1954698e596d728fc848e?
Maybe 3.11 nightly has the fix already?

**Comment by Dreamsorcerer** [comment]
It was backported the same day that 3.11.0 was released. So, I think the suspicion is that the "fix" may actually be causing the memory leak? Because it must have been in 3.11.1+, but probably just missed the 3.11.0 release, which aligns with the reports.

**Comment by Dreamsorcerer** [comment]
Going by https://github.com/elastic/rally/issues/1714, it looks like we just need to add some checks for None. Before that change, it would never be None, so the code worked correctly. I guess something about that exception could also end up resulting in a memory leak?

**Comment by bdraco** [comment]
I haven't seen the exception but I have a Home Assistant leaking about 450MiB of ram per day. I just turned off `enable_cleanup_closed` and should know tomorrow if the leak continues

**Comment by bdraco** [comment]
The leaked objects are `MemoryBIO`, `SSLProtocol`, `SSLObject`, `_SSLProtocolTransport` `memoryview`, and `managedbuffer`s

The `SSLProtocol` objects have no `socket` attached to them

The `asyncio.sslproto._SSLProtocolTransport` have no `SSLProtocol`  and no `socket`

Thats about as far as I've gotten with the debugging

**Comment by Dreamsorcerer** [comment]
Well, if you could also test with the changes in #7280 later, that'd be great.

**Comment by Dreamsorcerer** [comment]
I don't know if the exceptions might get suppressed for some reason, but I suppose that if an exception occurs on that abort(), then none of the other transports will get aborted, and none of the transports will get removed from that list in order to be garbage collected. So, it makes sense that a memory leak could occur.

**Comment by bdraco** [comment]
I'll switch my production instance to use https://github.com/aio-libs/aiohttp/pull/7280 after the `enable_cleanup_closed=False` test runs overnight

**Comment by bdraco** [comment]
Setting `enable_cleanup_closed=False` significant reduces the leak.  Not 100% sure if it means that we are trading the fast leak for a slow leak though.

Trying #7280 now.

**Comment by SishaarRao** [comment]
Hey all, seems like you all are triaging and attempting to solve the issue so best of luck 🙏 
Not sure if this is helpful, but I'm running `aiohttp==3.8.4` and Python version 3.11.2 in a production server. We have observed a steady increase in Inactive Memory(`proc/meminfo`) with a rate correlated with the rate HTTP requests. I haven't previously used the `enable_cleanup_closed` setting before however, so not sure where that leaves us as potentially being affected by this issue.

For the raw delta, we saw an increase in ~600MB in Inactive Memory over ~6.5k HTTP requests via `aiohttp`

**Comment by bdraco** [comment]
While its too soon to be sure (need to give it a few more hours), keeping `enable_cleanup_closed=True` along with https://github.com/aio-libs/aiohttp/pull/7280 does seem to have fixed the leak I am seeing.

**Comment by Dreamsorcerer** [comment]
> I haven't previously used the enable_cleanup_closed setting before however, so not sure where that leaves us as potentially being affected by this issue.

I think this change will only have an effect with `enable_cleanup_closed`. If there is a memory leak without that option, then it is likely a different issue. Worth testing with Python 3.11.0 to see if the memory leak is still there (which would rule out the same change being the cause).

**Comment by bdraco** [comment]
I can no longer replicate the issue after the change in #7280 

It would be great to get a 3.8.5 since this affects Home Assistant users in production.

**Comment by Dreamsorcerer** [comment]
I'm thinking it might be better to just get this fixed in cpython, 3.11.4 will be released in 3 weeks, so if we get it fixed there in the next week, it'll probably beat aiohttp to a new release anyway.

**Comment by Dreamsorcerer** [comment]
https://github.com/python/cpython/pull/104485

Should be fixed in 3.11.4. So, the only affected versions are 3.11.1, 3.11.2 and 3.11.3.

**Comment by bdraco** [comment]
Thanks 👍
